[PT2][COFFEE] pad mm + contigous layout change passes

mengluy0125 opened the pull request:
Summary:
context: https://docs.google.com/document/d/16i-vpuu4VsrKVsOZDq_VpLTfwoB4Ld8A3QHooWLsTTc/edit

1. Add two more passes, i.e., pad_mm + contigous layout enablement. We use build-in pad function, which shows faster in my experiments compared to pad_dim

2. Revive the large-k decompose with a separete pass to enable future use for other models in case we observe gains

Test Plan:
# local reproduce
checkout commit id: e9901bd610dd
set config as follows:
```
	
        torch._inductor.config.post_grad_fusion_options = {
            "make_mmt_contiguous_pass": {},
            "pad_mm_pass": {},
            "decompose_mm_pass": {},
        }
```
```
buck2 run mode/opt //scripts/weichenmeta/coffee:lce_perf
```
https://pxl.cl/4S0gf
P1361918604
see config setting and detailed results in D57312038

Differential Revision: D57197480




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

jackiexu1992 reviewed (approved):
Accept to unblock. Before land, please make sure it will not regress the current passes. Thanks

mengluy0125 committed:
[PT2][COFFEE] pad mm + contigous layout change passes (#126132)

Summary:

context: https://docs.google.com/document/d/16i-vpuu4VsrKVsOZDq_VpLTfwoB4Ld8A3QHooWLsTTc/edit

Add two more passes, i.e., pad_mm + contigous layout enablement. We use build-in pad function, which shows faster in my experiments compared to pad_dim

Test Plan:
# Sample config for APS to turn on the following opt. techs


```
dynamo_config_map:
    module_name_to_config:
      LinearCompressedEmbedding:
        backend: inductor
        cudagraph: false
        autotune: false
        dynamic_shapes: true
        skip_nnmodule_hook_guards: false
        cache_size_limit: 64
        optimize_ddp: false
        verbose_log: false
        post_grad_fusion_options:
          make_mmt_contiguous_pass: {}
          pad_mm_pass: {}
          decompose_mm_pass: {}

```


# local reproduce
checkout commit id: e9901bd610dd
set config as follows:
```

        torch._inductor.config.post_grad_fusion_options = {
            "make_mmt_contiguous_pass": {},
            "pad_mm_pass": {},
            "decompose_mm_pass": {},
        }
```
```
buck2 run mode/opt //scripts/weichenmeta/coffee:lce_perf
```
https://pxl.cl/4S0gf
P1361918604
see config setting and detailed results in D57312038

# e2e

### QPS
 {F1644045989}

###  NE

{F1644046461}{F1644046464}{F1644046463}

### trace

[link](https://www.internalfb.com/intern/perfdoctor/trace_view?filepath=tree%2Ftraces%2Fdynocli%2Faps-pt2_proposal_no_ad_impression_experimental_padding-bc3dfbd9fc%2F0%2Frank-1.May_19_23_39_13.3467.pt.trace.json.gz&bucket=aps_traces)

 {F1644061108}
 {F1644062141}

### summary

- QPS increased by > 6% by adding our passes
- with the current improved padding by eellison, we now see the padding itself actually helps to add in our passes. We can see from result, w/o padding, we can observe ~2.3% qps regression.
- NE seems neutral.
- We no longer observe long computation kernel in the backward.




### baseline w/o pt2
mast job: 
aps-pt2_pure_baseline_no_ad_impression_experimental-06547d6f59

 {F1644272801} 

~6.8 QPS regression by using PT2 + BF16

Reviewed By: jackiexu1992

Differential Revision: D57197480
